Fails to build on ESP32_GENERIC-SPIRAM due to iram overflow

Hi and thank you for this library.

I tried building it on ESP32, in order to run on a LilyGO TTGO T-Camera Mic ESP32. However, the build is failing due to running out of iram. I am using the lastest MicroPython release, v1.23.0 with esp-idf 5.0.4. Build is done with BOARD=ESP32_GENERIC BOARD_VARIANT=SPIRAM

Default build with mp_camera: `iram0_0_seg' overflowed by 1592 bytes
```
make USER_C_MODULES=../../../../mp_camera/src/micropython.cmake BOARD=ESP32_GENERIC BOARD_VARIANT=SPIRAM all

/home/jon/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: micropython.elf section `.iram0.text' will not fit in region `iram0_0_seg'
/home/jon/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: IRAM0 segment data does not fit.
/home/jon/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: region `iram0_0_seg' overflowed by 1592 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```

Default build without mp_camera - works. Unsure how many bytes are left in iram...
```
make BOARD=ESP32_GENERIC BOARD_VARIANT=SPIRAM all

micropython.bin binary size 0x17a5f0 bytes. Smallest app partition is 0x1f0000 bytes. 0x75a10 bytes (24%) free.
...
bootloader  @0x001000    23072  (    5600 remaining)
partitions  @0x008000     3072  (    1024 remaining)
application @0x010000  1549808  (  481808 remaining)
```

I tried to disable Ethernet support to reduce iram usage. A tip from a similar problem with iram overflow at https://github.com/orgs/micropython/discussions/14263#discussioncomment-9047807 

Using sdkconfig.base of
```
CONFIG_ETH_USE_SPI_ETHERNET=n
CONFIG_ETH_SPI_ETHERNET_W5500=n
CONFIG_ETH_SPI_ETHERNET_KSZ8851SNL=n
CONFIG_ETH_SPI_ETHERNET_DM9051=n
```

Still runs out of iram, just by 500 bytes less:
```
sram-cache-issue  -mfix-esp32-psram-cache-strategy=memw && :
/home/jon/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: micropython.elf section `.iram0.text' will not fit in region `iram0_0_seg'
/home/jon/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: IRAM0 segment data does not fit.
/home/jon/.espressif/tools/xtensa-esp32-elf/esp-2022r1-11.2.0/xtensa-esp32-elf/bin/../lib/gcc/xtensa-esp32-elf/11.2.0/../../../../xtensa-esp32-elf/bin/ld: region `iram0_0_seg' overflowed by 1024 bytes
collect2: error: ld returned 1 exit status
ninja: build stopped: subcommand failed.
```
I am not sure how to reduce iram usage further in the ESP32 build. Any ideas? Or perhaps there is something this module can do differently to reduce iram usage?

--- a/.github/workflows/ESP32.yml
+++ b/.github/workflows/ESP32.yml
@@ -3,6 +3,8 @@ name: Build MicroPython Camera for ESP32 Boards
 on:
   workflow_dispatch:
   push:
+    paths:
+      - 'src/**'
   pull_request:
     branches:
       - master

--- a/.github/workflows/ESP32.yml
+++ b/.github/workflows/ESP32.yml
@@ -140,4 +140,4 @@ jobs:
         with:
           name: firmware-${{ matrix.board }}
           path: ~/${{ matrix.board }}.bin
-          retention-days: 1
+          retention-days: 90

--- a/.github/workflows/ESP32.yml
+++ b/.github/workflows/ESP32.yml
@@ -1,4 +1,4 @@
-name: Build MicroPython Camera for ESP32 Boards
+name: ESP32
 
 on:
   workflow_dispatch:

--- a/README.md
+++ b/README.md
@@ -3,8 +3,8 @@ This project aims to support cameras in different ports in micropython, starting
 At the moment, this is a micropython user module, but it might get in the micropython repo in the future.
 The API is stable, but it might change without previous anounce.
 
-## Precomiled FW (the easy way)
-If you are not familiar with building a custom firmware, you can go to the actions tab and start the build workflow by yourself. Then, you can download one if the generic FWs that suits your board.
+## Precomiled FW (the easy way) (work in progress)
+If you are not familiar with building a custom firmware, you can go to the actions tab and start the build workflow by yourself. Then, you can download one of the generic FWs that suits your board.
 
 ## Setup build environment (the DIY way)
 To build the project, follow the following instructions:

--- a/README.md
+++ b/README.md
@@ -8,7 +8,7 @@ If you are not familiar with building a custom firmware, you can go to the actio
 
 ## Setup build environment (the DIY way)
 To build the project, follow the following instructions:
-- [ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/v5.2.2/esp32/get-started/index.html): I used version 5.2.2, but it might work with other versions.
+- [ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/v5.2.3/esp32/get-started/index.html): I used version 5.2.2, but it might work with other versions (see notes).
 - Clone the micropython repo and this repo in a folder, e.g. "MyESPCam". I used the actual micropython master branch (between v1.23 and before 1.24).
 - You will have to add the ESP32-Camera driver (I used v2.0.12). To do this, add the following to the respective idf_component.yml file  (e.g. in micropython/ports/esp32/main_esp32s3/idf_component.yml):
 ```
@@ -95,3 +95,6 @@ camera.set_quality(10)
 You can get and set sensor properties by the respective methods (e.g. camera.get_brightness() or camera.set_vflip(True). See autocompletitions in Thonny in order to see the list of methods.
 If you want more insides in the methods and what they actually do, you can find a very good documentation [here](https://docs.circuitpython.org/en/latest/shared-bindings/espcamera/index.html).
 Notice that for the methods in here you need to prefix a get/set, depending that you want to do.
+
+## Notes
+If your target board is a ESP32, I recomend using IDF >= 5.2, since older versions may lead to IRAM overflow during build. Alternatively you can modify your sdkconfig-file (see #1).

--- a/README.md
+++ b/README.md
@@ -97,4 +97,4 @@ If you want more insides in the methods and what they actually do, you can find 
 Notice that for the methods in here you need to prefix a get/set, depending that you want to do.
 
 ## Notes
-If your target board is a ESP32, I recomend using IDF >= 5.2, since older versions may lead to IRAM overflow during build. Alternatively you can modify your sdkconfig-file (see #1).
+If your target board is a ESP32, I recomend using IDF >= 5.2, since older versions may lead to IRAM overflow during build. Alternatively you can modify your sdkconfig-file (see [issue #1](https://github.com/cnadler86/micropython-camera-API/issues/1)).

--- a/README.md
+++ b/README.md
@@ -1,4 +1,4 @@
-# Camera API for micropython.
+# Camera API for micropython
 [![ESP32 Port](https://github.com/cnadler86/micropython-camera-API/actions/workflows/ESP32.yml/badge.svg)](https://github.com/cnadler86/micropython-camera-API/actions/workflows/ESP32.yml)
 
 This project aims to support cameras in different ports in micropython, starting with the ESP32-Port and omnivision (OV2640 & OV5640) cameras. The project implements a general API for cameras in micropython (such as circuitpython have done).

--- a/README.md
+++ b/README.md
@@ -1,12 +1,14 @@
 # Camera API for micropython.
+[![ESP32 Port](https://github.com/cnadler86/micropython-camera-API/actions/workflows/ESP32.yml/badge.svg)](https://github.com/cnadler86/micropython-camera-API/actions/workflows/ESP32.yml)
 This project aims to support cameras in different ports in micropython, starting with the ESP32-Port and omnivision (OV2640 & OV5640) cameras. The project implements a general API for cameras in micropython (such as circuitpython have done).
 At the moment, this is a micropython user module, but it might get in the micropython repo in the future.
 The API is stable, but it might change without previous anounce.
 
 ## Precomiled FW (the easy way) (work in progress)
-If you are not familiar with building a custom firmware, you can go to the actions tab and start the build workflow by yourself. Then, you can download one of the generic FWs that suits your board.
+If you are not familiar with building a custom firmware, you can [click here](https://nightly.link/cnadler86/micropython-camera-API/workflows/ESP32/master?preview) and download one of the generic FWs that suits your board.
 
-## Setup build environment (the DIY way)
+## Build your custom FW
+### Setup build environment (the DIY way)
 To build the project, follow the following instructions:
 - [ESP-IDF](https://docs.espressif.com/projects/esp-idf/en/v5.2.3/esp32/get-started/index.html): I used version 5.2.2, but it might work with other versions (see notes).
 - Clone the micropython repo and this repo in a folder, e.g. "MyESPCam". I used the actual micropython master branch (between v1.23 and before 1.24).
@@ -17,7 +19,7 @@ To build the project, follow the following instructions:
 ```
 You can also clone the https://github.com/espressif/esp32-camera repository inside the esp-idf/components folder instead of altering the idf_component.yml file.
 
-## Add camera configurations to your board (Optional, but recomended)
+### Add camera configurations to your board (Optional, but recomended)
 To make things easier, add the following lines to your board config-file "mpconfigboard.h" with the respective pins and camera parameters. Otherwise you will need to pass all parameters during construction.
 Don't forget the empty line at the buttom.
 Example for xiao sense:
@@ -46,7 +48,7 @@ Example for xiao sense:
 
 ```
 
-## Build the API
+### Build the API
 To build the project, you could do it the following way:
 
 ```bash

--- a/README.md
+++ b/README.md
@@ -5,8 +5,46 @@ This project aims to support cameras in different ports in micropython, starting
 At the moment, this is a micropython user module, but it might get in the micropython repo in the future.
 The API is stable, but it might change without previous anounce.
 
-## Precomiled FW (the easy way) (work in progress)
-If you are not familiar with building a custom firmware, you can [click here](https://nightly.link/cnadler86/micropython-camera-API/workflows/ESP32/master?preview) and download one of the generic FWs that suits your board.
+## Precomiled FW (the easy way)
+If you are not familiar with building a custom firmware, you can go to the [releases](https://github.com/cnadler86/micropython-camera-API/releases) page and download one of the generic FWs that suits your board.
+
+## Using the API
+```python
+from camera import Camera, GrabMode, PixelFormat, FrameSize, GainCeiling
+
+# Camera construction and initialization
+camera = Camera(
+    data_pins=[1,2,3,4,5,6,7,8],
+    vsync_pin=9,
+    href_pin=10,
+    sda_pin=11,
+    scl_pin=12,
+    pclk_pin=13,
+    xclk_pin=14,
+    xclk_freq=20000000,
+    powerdown_pin=-1,
+    reset_pin=-1,
+    pixel_format=PixelFormat.RGB565,
+    frame_size=FrameSize.QVGA,
+    jpeg_quality=15,
+    fb_count=1,
+    grab_mode=GrabMode.WHEN_EMPTY
+)
+
+#Camera construction using defaults (if you specified them in mpconfigboard.h)
+camera = Camera()
+
+# Capture image
+img = camera.capture()
+
+# Camera reconfiguration 
+camera.reconfigure(pixel_format=PixelFormat.JPEG,frame_size=FrameSize.QVGA,grab_mode=GrabMode.LATEST, fb_count=2)
+camera.set_quality(10)
+```
+
+You can get and set sensor properties by the respective methods (e.g. camera.get_brightness() or camera.set_vflip(True). See autocompletitions in Thonny in order to see the list of methods.
+If you want more insides in the methods and what they actually do, you can find a very good documentation [here](https://docs.circuitpython.org/en/latest/shared-bindings/espcamera/index.html).
+Notice that for the methods in here you need to prefix a get/set, depending that you want to do.
 
 ## Build your custom FW
 ### Setup build environment (the DIY way)
@@ -55,49 +93,11 @@ To build the project, you could do it the following way:
 ```bash
 $ . <path2esp-idf>/esp-idf/export.sh
 $ cd MyESPCam/micropython/ports/esp32
-$ make USER_C_MODULES=../../../../mp_camera/src/micropython.cmake BOARD=<Your-Board> clean
-$ make USER_C_MODULES=../../../../mp_camera/src/micropython.cmake BOARD=<Your-Board> submodules
-$ make USER_C_MODULES=../../../../mp_camera/src/micropython.cmake BOARD=<Your-Board> all
+$ make USER_C_MODULES=../../../../micropython-camera-API/src/micropython.cmake BOARD=<Your-Board> clean
+$ make USER_C_MODULES=../../../../micropython-camera-API/src/micropython.cmake BOARD=<Your-Board> submodules
+$ make USER_C_MODULES=../../../../micropython-camera-API/src/micropython.cmake BOARD=<Your-Board> all
 ```
 if you experience problems, visit [MicroPython external C modules](https://docs.micropython.org/en/latest/develop/cmodules.html).
-
-## Using the API
-```python
-from camera import Camera, GrabMode, PixelFormat, FrameSize, GainCeiling
-
-# Camera construction and initialization
-camera = Camera(
-    data_pins=[1,2,3,4,5,6,7,8],
-    vsync_pin=9,
-    href_pin=10,
-    sda_pin=11,
-    scl_pin=12,
-    pclk_pin=13,
-    xclk_pin=14,
-    xclk_freq=20000000,
-    powerdown_pin=-1,
-    reset_pin=-1,
-    pixel_format=PixelFormat.RGB565,
-    frame_size=FrameSize.QVGA,
-    jpeg_quality=15,
-    fb_count=1,
-    grab_mode=GrabMode.WHEN_EMPTY
-)
-
-#Camera construction using defaults (if you specified them in mpconfigboard.h)
-camera = Camera()
-
-# Capture image
-img = camera.capture()
-
-# Camera reconfiguration 
-camera.reconfigure(pixel_format=PixelFormat.JPEG,frame_size=FrameSize.QVGA,grab_mode=GrabMode.LATEST, fb_count=2)
-camera.set_quality(10)
-```
-
-You can get and set sensor properties by the respective methods (e.g. camera.get_brightness() or camera.set_vflip(True). See autocompletitions in Thonny in order to see the list of methods.
-If you want more insides in the methods and what they actually do, you can find a very good documentation [here](https://docs.circuitpython.org/en/latest/shared-bindings/espcamera/index.html).
-Notice that for the methods in here you need to prefix a get/set, depending that you want to do.
 
 ## Notes
 If your target board is a ESP32, I recomend using IDF >= 5.2, since older versions may lead to IRAM overflow during build. Alternatively you can modify your sdkconfig-file (see [issue #1](https://github.com/cnadler86/micropython-camera-API/issues/1)).

--- a/README.md
+++ b/README.md
@@ -1,5 +1,6 @@
 # Camera API for micropython.
 [![ESP32 Port](https://github.com/cnadler86/micropython-camera-API/actions/workflows/ESP32.yml/badge.svg)](https://github.com/cnadler86/micropython-camera-API/actions/workflows/ESP32.yml)
+
 This project aims to support cameras in different ports in micropython, starting with the ESP32-Port and omnivision (OV2640 & OV5640) cameras. The project implements a general API for cameras in micropython (such as circuitpython have done).
 At the moment, this is a micropython user module, but it might get in the micropython repo in the future.
 The API is stable, but it might change without previous anounce.

--- a/src/modcamera.c
+++ b/src/modcamera.c
@@ -35,30 +35,29 @@
 #error Camera only works on boards configured with spiram
 #endif
 
-//TODO: improve error message
 void raise_micropython_error_from_esp_err(esp_err_t err) {
     switch (err) {
         case ESP_OK:
             return;
 
         case ESP_ERR_NO_MEM:
-            mp_raise_msg(&mp_type_MemoryError, MP_ERROR_TEXT("ESP_ERR_NO_MEM: Out of memory"));
+            mp_raise_msg(&mp_type_MemoryError, MP_ERROR_TEXT("Out of memory"));
             break;
 
         case ESP_ERR_INVALID_ARG:
-            mp_raise_ValueError(MP_ERROR_TEXT("ESP_ERR_INVALID_ARG: Invalid argument"));
+            mp_raise_ValueError(MP_ERROR_TEXT("Invalid argument"));
             break;
 
         case ESP_ERR_INVALID_STATE:
-            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("ESP_ERR_INVALID_STATE: Invalid state"));
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Invalid state"));
             break;
 
         case ESP_ERR_NOT_FOUND:
-            mp_raise_OSError(MP_ENOENT);
+            mp_raise_msg(&mp_type_OSError, MP_ERROR_TEXT("Camera not found"));
             break;
 
         case ESP_ERR_NOT_SUPPORTED:
-            mp_raise_NotImplementedError(MP_ERROR_TEXT("ESP_ERR_NOT_SUPPORTED: Operation not supported"));
+            mp_raise_NotImplementedError(MP_ERROR_TEXT("Operation/Function not supported/implemented"));
             break;
 
         case ESP_ERR_TIMEOUT:
@@ -66,8 +65,8 @@ void raise_micropython_error_from_esp_err(esp_err_t err) {
             break;
 
         default:
-            // mp_raise_msg_varg(&mp_type_RuntimeError, MP_ERROR_TEXT("ESP_ERR: Unknown error 0x%04x"), err);
-            mp_raise_msg_varg(&mp_type_RuntimeError, MP_ERROR_TEXT("ESP_ERR: Unknown error"));
+            mp_raise_msg_varg(&mp_type_RuntimeError, MP_ERROR_TEXT("Unknown error 0x%04x"), err);
+            // mp_raise_msg_varg(&mp_type_RuntimeError, MP_ERROR_TEXT("Unknown error"));
             break;
     }
 }

--- a/src/modcamera.c
+++ b/src/modcamera.c
@@ -31,7 +31,7 @@
 
 #define TAG "ESP32_MPY_CAMERA"
 
-#if !CONFIG_SPIRAM //TODO: Better test if enought RAM is available on runtime?
+#if !CONFIG_SPIRAM
 #error Camera only works on boards configured with spiram
 #endif
 
@@ -150,6 +150,7 @@ void mp_camera_hal_deinit(mp_camera_obj_t *self) {
         esp_err_t err = esp_camera_deinit();
         raise_micropython_error_from_esp_err(err);
         self->initialized = false;
+        ESP_LOGI(TAG, "Camera deinitialized");
     }
 }
 
@@ -289,7 +290,7 @@ const mp_rom_map_elem_t mp_camera_hal_gainceiling_table[] = {
     { MP_ROM_QSTR(MP_QSTR_128X),    MP_ROM_INT(GAINCEILING_128X) },
 };
 
-//OPEN: Makros with convertion function, since the API will use standarized values.
+//TODO: Makros with convertion function, since the API will use standarized values.
 // Helper functions to get and set camera and sensor information
 #define SENSOR_STATUS_GETSET_IN_RANGE(type, name, status_field_name, setter_function_name, min_val, max_val) \
     SENSOR_GETSET_IN_RANGE(type, name, status.status_field_name, setter_function_name, min_val, max_val)

